lib/plugins/attributeAlias.js

```diff
@@ -74,6 +78,29 @@ function applyAlias(mappings) {
             attribute.name = mappings.direct[attribute.name];
         }
 
+        if (config.checkNgsi2() && typeof(attribute.value) === 'string') {
```

Comment:
Doubt: in the case of IOTA JSON (which, given the JSON syntax, allows to specify the type directly in the measures sent by the devide), what takes preference? The type comming at the measure? Or the one in `attribute.type`?

Let me illustrate with an example. Let's consider we have provising this way:

```
      "attributes": [
        {
          "name": "temperature",
          "object_id": "t",
          "type": "Integer"
        }
      ]
```

Case 1 (without alias): what is progressed to CB in the case the device send this measure? `23` or `"23"`?

```
{
  "temperature": "23"
}
```

Case 2 (with alias): what is progressed to CB in the case the device send this measure? `23` or `"23"`?

```
{
  "t": "23"
}
```

Reply:
With 6d965a5e19e84820433154c8486652366d60cdae, I have added a couple of additional tests to reflect you doubt. Are they correct?

Reply:
Nice... to use an unit test to answer a question is the best way to do it! :)

Thus, as far as I understand, the cast done by the library overrides the type used in the southubound in both cases (i.e. using alias or not using it doesn't affect this behaviour). 

Not sure if this the best behaviour... I mean, casting to number, boolean, etc. is a nice feature to overpass limitations on the southbound API, i.e. UL protocol which is strictly string-based (not sure about LWM2M). However, it the southbound API syntax is expressive enough and doesn't have such limitation I understand it should take preference over the casting.

Having said that, I'm not sure how the library could enable/disable casting depending on the capabilities of the southbound API.

What do you think?

Reply:
On a second though... maybe it should be a configurable aspect (i.e. config.js). Thus, the user of the agent can choose how autocast is used (i.e. IOTA-UL users will probably activate it, IOTA-JSON users will probably disable it).

Reply:
After your comments, I have been thinking and I discovered that the implemented functionality was not completely correct. I have modified it with 8148d99a0c9bf075cbfa6f0e8448a4dfdf5bcb95.

To summarize:

1. Case 1 (without alias): what is progressed to CB in the case the device send this measure? 23 or "23"?

```
{
  "temperature": "23"
}
```

The IOTa is reporting the observation without using the alias. Therefore the plugin functionality should not be applied and the value should be sent to the CB as it was received in the southbound.

Case 2 (with alias): what is progressed to CB in the case the device send this measure? 23 or "23"?
```
{
  "t": "23"
}
```

The IOTa uses the alias and therefore the casting functionality of the plugin is applied and the value is sent to the CB with the corresponding JSON native type.

Now, I feel that we have a correct functionality.

Reply:
Regarding one of your comments:

> However, it the southbound API syntax is expressive enough and doesn't have such limitation I understand it should take preference over the casting.

Note that in https://github.com/dcalvoalonso/iotagent-node-lib/blob/task/aliasPluginNgsi2/lib/plugins/attributeAlias.js#L80, we only applied the casting if we reported attributes are received in String format. Therefore, if the observations are received in the southbound API of the IOTa using the appropriate format (i.e., integer, float...), this is not applied.

Reply:
With regards to "correct functionality", under my understanding, the alias plugin is just a way of having short names (i.e. t, h, p, etc.) in the southbound interface. From: https://github.com/telefonicaid/iotagent-node-lib#attribute-alias-plugin-attributealias

>Attribute Alias plugin (attributeAlias)
>
> In the Device provision, an id can be specified for each attribute, along with its name. The Id can be used then as the left part of a mapping from attribute names in the south bound to attribute names in the North Bound. If the id and name attributes are used in this way, this plugin makes the translation from one to the other automatically.

Thus, I understand that the behavoir with regards to casting should be the same in both cases (i.e. direct measures without using the alias plugin and measures using it). But maybe I'm missing something... Do you think it should be different? Why?

With regards to https://github.com/dcalvoalonso/iotagent-node-lib/blob/task/aliasPluginNgsi2/lib/plugins/attributeAlias.js#L80, I understand your point (and it is fine for most of the cases :), but it is a bit weird (from a user perspetive) to send `{ "temperature": "23" }` at southbound and get `{ "temperature": 23}` at CB... becase if me (as an user) want to have `{ "temperature": 23}` at CB then I would have sent `{ {"temperature": 23}` in the southbound, given that JSON format allows me to use directly numbers.

Making this behaviour configurable, e.g something like this

```
if (config.checkNgsi2() && config.checkSbStringCast() && typeof(attribute.value) === 'string') 
```

would avoid this kind of problems, don't you think?

Reply:
So, if understand your proposal correctly, you suggest increasing the scope of the PR so that the casting from string to the corresponding native type shall be done not only for attributes reported with alias but for any attribute, right?

Reply:
Yes, that's correct. In the case the fix would be easy, of course (maybe it is a matter of moving the same code to another point of the library...).

Reply:
I have done the first attempt to implement this with 19708645f3af9feff89dbdd05f41d04df3e6b3ef . @fgalan , please check what do you think about it!

Reply:
It looks good. I understand the new function `castJsonNativeAttributes()` is applied in general to measures, no matter if they used alias plugin or not.

We can close this comment thread. I'll provide more specific feedback on the new implementation on serparated comment.

NTC